### PR TITLE
TMA-680: Add option include_computed_attributes

### PR DESCRIPTION
**Pull request by panjan**

### CODE REVIEW

File: spec/unit/actions/synchronize_cas_spec.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        synchronize: synchronize,
+        gdc_logger: logger
+    }
+  end
+  let(:params) { basic_params }
+  let(:converted_params) do
+    GoodData::LCM2.convert_to_smart_hash(params)
+  end
+
+  before do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_cas_spec.rb]: Identical code found in synchronize_cas_spec.rb:27, synchronize_ldm_spec.rb:27 (mass = 100)

File: spec/unit/actions/synchronize_ldm_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        synchronize: synchronize,
+        gdc_logger: logger
+    }
+  end
+  let(:params) { basic_params }
+  let(:converted_params) do
+    GoodData::LCM2.convert_to_smart_hash(params)
+  end
+
+  before do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_ldm_spec.rb]: Identical code found in synchronize_cas_spec.rb:27, synchronize_ldm_spec.rb:27 (mass = 100)

File: spec/unit/actions/synchronize_cas_spec.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    allow(gdc_gd_client).to receive(:projects).and_return(project)
+    allow(project).to receive(:title)
+    allow(project).to receive(:pid)
+    allow(logger).to receive(:info)
+  end
+
+  context 'when include_computed_attributes is not specified' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_cas_spec.rb]: Similar code found in synchronize_cas_spec.rb:34, synchronize_cas_spec.rb:43 (mass = 64)

File: spec/unit/actions/synchronize_cas_spec.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    let(:params) { basic_params.merge(include_computed_attributes: 'true') }
+    it 'it synchronizes computed attributes' do
+      expect(project).to receive(:execute_maql).with('expected_maql')
+      result = subject.class.call(converted_params)
+      expect(result).not_to be_empty
+    end
+  end
+
+  context 'when include_computed_attributes is true' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_cas_spec.rb]: Similar code found in synchronize_cas_spec.rb:34, synchronize_cas_spec.rb:43 (mass = 64)

File: spec/unit/actions/synchronize_ldm_spec.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    allow(gdc_gd_client).to receive(:projects).and_return(project)
+    allow(project).to receive(:title)
+    allow(project).to receive(:pid)
+    allow(logger).to receive(:info)
+  end
+
+  context 'when include_computed_attributes is not specified' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_ldm_spec.rb]: Similar code found in synchronize_ldm_spec.rb:34, synchronize_ldm_spec.rb:42 (mass = 58)

File: spec/unit/actions/synchronize_ldm_spec.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    let(:params) { basic_params.merge(include_computed_attributes: 'true') }
+    it 'diffs LDM with computed attributes' do
+      expect(project).to receive(:blueprint).with(include_ca: true)
+      subject.class.call(converted_params)
+    end
+  end
+
+  context 'when include_computed_attributes is true' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_ldm_spec.rb]: Similar code found in synchronize_ldm_spec.rb:34, synchronize_ldm_spec.rb:42 (mass = 58)

File: spec/unit/models/project_spec.rb [hunk 7]
```diff
@@ -40,4 +40,51 @@
...
+    subject { GoodData::Project.new(project_data) }
+    before do
+      allow(client).to receive(:get).and_return(diff_result)
+      allow(client).to receive(:poll_on_code)
+      allow(GoodData::Model::FromWire).to receive(:from_wire) .and_return(blueprint)
+      allow(blueprint).to receive(:title=)
+      subject.client = client
+    end
+
+    context 'when include_ca option is true' do
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Similar code found in project_spec.rb:63, project_spec.rb:81 (mass = 52)

File: spec/unit/models/project_spec.rb [hunk 8]
```diff
@@ -40,4 +40,51 @@
...
+    context 'when include_ca option is false' do
+      it 'requests model view with includeCA=false' do
+        expect(client).to receive(:get).with('/gdc/projects/foo/model/view',
+                           hash_including(params:
+                             hash_including(includeCA: false)))
+        subject.blueprint(include_ca: false)
+      end
+    end
+
+    context 'when include_ca option is not specified' do
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Similar code found in project_spec.rb:63, project_spec.rb:81 (mass = 52)

File: lib/gooddata/lcm/actions/synchronize_cas.rb [hunk 9]
```diff
@@ -17,11 +17,22 @@ class SynchronizeComputedAttributes < BaseAction
...
+
+        description 'Specifies whether to transfer computed attributes'
+        param :include_computed_attributes, instance_of(Type::BooleanType), required: false, default: true
       end
 
       class << self
         def call(params)
+          # set default value for include_computed_attributes
+          # (we won't have to do this after TMA-690)
+          include_ca = params.include_computed_attributes
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_cas.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_cas.rb [hunk 10]
```diff
@@ -17,11 +17,22 @@ class SynchronizeComputedAttributes < BaseAction
...
+          include_ca = true if include_ca.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_cas.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/actions/synchronize_ldm.rb [hunk 11]
```diff
@@ -23,10 +23,19 @@ class SynchronizeLdm < BaseAction
...
+
+        description 'Specifies whether to transfer computed attributes'
+        param :include_computed_attributes, instance_of(Type::BooleanType), required: false, default: true
       end
 
       class << self
         def call(params)
+          # set default value for include_computed_attributes
+          # (we won't have to do this after TMA-690)
+          include_ca = params.include_computed_attributes
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_ldm.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_ldm.rb [hunk 12]
```diff
@@ -23,10 +23,19 @@ class SynchronizeLdm < BaseAction
...
+          include_ca = true if include_ca.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_ldm.rb]: Performs a nil-check (NilCheck)

File: spec/unit/actions/synchronize_cas_spec.rb [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+describe GoodData::LCM2::SynchronizeComputedAttributes do
+  let(:gdc_gd_client) { double(GoodData::Rest::Client) }
+  let(:logger) { double(Logger) }
+  let(:project) { double(GoodData::Project) }
+  let(:synchronize) do
+    [{ from: 'foo', to: [{ ca_scripts: { 'maqlDdlChunks' => ['expected_maql'] } }] }]
+  end
+  let(:basic_params) do
+    {
+        gdc_gd_client: gdc_gd_client,
```
Comment (yenkins) [inline spec/unit/actions/synchronize_cas_spec.rb]: Style/IndentHash: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.

File: spec/unit/actions/synchronize_ldm_spec.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+describe GoodData::LCM2::SynchronizeLdm do
+  let(:gdc_gd_client) { double(GoodData::Rest::Client) }
+  let(:logger) { double(Logger) }
+  let(:project) { double(GoodData::Project) }
+  let(:synchronize) do
+    [{ from: 'foo', to: [] }]
+  end
+  let(:basic_params) do
+    {
+        gdc_gd_client: gdc_gd_client,
```
Comment (yenkins) [inline spec/unit/actions/synchronize_ldm_spec.rb]: Style/IndentHash: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.

File: spec/unit/models/project_spec.rb [hunk 15]
```diff
@@ -40,4 +40,51 @@
       end
     end
   end
+
+  describe '#blueprint' do
+    let(:project_data) do
+      { 'project' => {
+          'links' => { 'self' => '/gdc/foo' },
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Style/IndentHash: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.

File: spec/unit/models/project_spec.rb [hunk 16]
```diff
@@ -40,4 +40,51 @@
...
+          'meta' => { 'title' => 'My Project' } } }
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Style/MultilineHashBraceLayout: Closing hash brace must be on the line after the last hash element when opening brace is on a separate line from the first hash element.

File: spec/unit/models/project_spec.rb [hunk 17]
```diff
@@ -40,4 +40,51 @@
...
+      allow(client).to receive(:poll_on_code)
+      allow(GoodData::Model::FromWire).to receive(:from_wire) .and_return(blueprint)
+      allow(blueprint).to receive(:title=)
+      subject.client = client
+    end
+
+    context 'when include_ca option is true' do
+      it 'requests model view with includeCA=true' do
+        expect(client).to receive(:get).with('/gdc/projects/foo/model/view',
+                           hash_including(params:
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Style/AlignParameters: Align the parameters of a method call if they span more than one line.

File: spec/unit/models/project_spec.rb [hunk 18]
```diff
@@ -40,4 +40,51 @@
...
+                             hash_including(includeCA: true)))
+        subject.blueprint(include_ca: true)
+      end
+    end
+
+    context 'when include_ca option is false' do
+      it 'requests model view with includeCA=false' do
+        expect(client).to receive(:get).with('/gdc/projects/foo/model/view',
+                           hash_including(params:
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Style/AlignParameters: Align the parameters of a method call if they span more than one line.

File: spec/unit/models/project_spec.rb [hunk 19]
```diff
@@ -40,4 +40,51 @@
...
+                             hash_including(includeCA: false)))
+        subject.blueprint(include_ca: false)
+      end
+    end
+
+    context 'when include_ca option is not specified' do
+      it 'requests model view with includeCA=true' do
+        expect(client).to receive(:get).with('/gdc/projects/foo/model/view',
+                           hash_including(params:
```
Comment (yenkins) [inline spec/unit/models/project_spec.rb]: Style/AlignParameters: Align the parameters of a method call if they span more than one line.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/535/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/288/ : FAILURE in 1m 58s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/595/ : SUCCESS in 1m 41s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/536/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1089/ : SUCCESS in 46m 36s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/848/ : FAILURE in 31s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/540/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/293/ : SUCCESS in 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/600/ : SUCCESS in 32s

File: spec/unit/actions/synchronize_cas_spec.rb [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+  before do
+    allow(gdc_gd_client).to receive(:projects).and_return(project)
+    allow(project).to receive(:title)
+    allow(project).to receive(:pid)
+    allow(logger).to receive(:info)
+  end
+
+  context 'when include_computed_attributes is not specified' do
+    let(:params) { basic_params.merge(include_computed_attributes: 'true') }
```
Comment (kubamahnert) [inline spec/unit/actions/synchronize_cas_spec.rb]: isn't this test the same as 'when include_computed_attributes is true' ?

**Comment by kubamahnert** [comment]
ok to merge

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/543/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/303/ : SUCCESS in 2m 05s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/610/ : SUCCESS in 2m 00s

**Comment by kubamahnert** [comment]
ok to merge

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/203/ : SUCCESS in 16s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/311/ : SUCCESS in 36s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/618/ : SUCCESS in 34s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1105/ : FAILURE in 48m 15s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/861/ : UNSTABLE in 6m 26s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/205/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/315/ : SUCCESS in 1m 48s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/622/ : SUCCESS in 1m 46s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1106/ : FAILURE in 46m 58s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/862/ : FAILURE in 38s

**Comment by panjan** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/549/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/317/ : SUCCESS in 36s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/624/ : SUCCESS in 34s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/207/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/318/ : SUCCESS in 33s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/625/ : SUCCESS in 35s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1108/ : FAILURE in 46m 05s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/864/ : UNSTABLE in 6m 48s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/210/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/321/ : SUCCESS in 51s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/628/ : SUCCESS in 35s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1109/ : FAILURE in 46m 22s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/865/ : UNSTABLE in 22m 48s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/226/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/342/ : SUCCESS in 33s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/649/ : SUCCESS in 28s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1123/ : SUCCESS in 46m 23s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/879/ : SUCCESS in 22m 31s

**Pull request merged**